core/cache/controller.go

```diff
@@ -119,6 +128,26 @@ func (c *Controller) loop() error {
 	}
 }
 
+// Mark updates all cached entities to indicate they are stale.
+func (c *Controller) Mark() {
+	c.marked = c.manager.mark()
+}
+
+// Marked indicates that there have been entities in the cache marked stale,
+// and a sweep is required.
+func (c *Controller) Marked() bool {
+	return c.marked
+}
+
+// Sweep evicts any stale entities from the cache,
+// cleaning up resources that they are responsible for.
+func (c *Controller) Sweep() {
+	if c.marked {
+		<-c.manager.sweep()
```

Comment:
Should there be a timeout here, or do we just block indefinitely?

Reply:
I did have a time-out upstream, but @howbazaar advised to thread the `tomb.Dying` through instead. 

Got no issue with time-out, but I think there needs to be a clear course of action if it is hit.

Reply:
yeah, a bare channel read was where I was mainly concerned. Any change of a comment that states that tomb.Dying is used instead of a timeout.

Also when I wrote mine, I wasn't entirely sure how to handle a error from a sweep, so the fact that it doesn't return makes it more cleaner from the watcher side.

Reply:
So a bare read has the problem that we can't guarantee it will ever return. So if one side shuts down, and never sends a message, then whatever called Sweep is permanently blocked.

I'm not particularly concerned about "how long a Sweep will take", as much as I am concerned about "during shutdown, one thread starts to call Sweep at exact the the time the other thread decides to stop and not do the work anymore".